storage/memory/storage.go

```diff
@@ -0,0 +1,268 @@
+		}
+	}
+	return store
+}
+
+var _ storage.BlockIndexer = &BlockStorage{}
+
+type BlockStorage struct {
+	base *baseStorage
+}
+
+func NewBlockStorage() *BlockStorage {
+	return &BlockStorage{
+		base: baseStorageFactory(),
+	}
+}
+
+func (s BlockStorage) GetByHeight(height uint64) (*types.Block, error) {
+	s.base.mu.RLock()
+	defer s.base.mu.RUnlock()
+
+	// Check if the requested height is within the known range
+	if height < s.base.firstHeight || height > s.base.lastHeight {
+		return nil, errors.NotFound
+	}
+
+	// Retrieve the block using the blockHeightsIDs map
+	blockID, exists := s.base.blockHeightsIDs[height]
+	if !exists {
+		return nil, errors.NotFound
+	}
+
+	// Retrieve the block using the blocksIDs map
+	block, exists := s.base.blocksIDs[blockID]
+	if !exists {
+		return nil, errors.NotFound
+	}
+
+	return block, nil
+}
+
+func (s BlockStorage) GetByID(ID common.Hash) (*types.Block, error) {
+	s.base.mu.RLock()
+	defer s.base.mu.RUnlock()
+
+	// Retrieve the block using the blocksIDs map
+	block, exists := s.base.blocksIDs[ID]
+	if !exists {
+		return nil, errors.NotFound
+	}
+
+	return block, nil
+}
+
+func (s BlockStorage) Store(block *types.Block) error {
+	s.base.mu.Lock()
+	defer s.base.mu.Unlock()
+
+	ID, err := block.Hash()
+	if err != nil {
+		return fmt.Errorf("block hash error: %w", err)
+	}
+
+	// Check if the block already exists
+	_, exists := s.base.blocksIDs[ID]
+	if exists {
+		return errors.Duplicate
+	}
+
+	// Store the block in blocksIDs map and update blockHeightsIDs map
+	s.base.blocksIDs[ID] = block
+	s.base.blockHeightsIDs[block.Height] = ID
+
+	// Update firstHeight and lastHeight if necessary
+	if s.base.firstHeight == unknownHeight {
+		s.base.firstHeight = block.Height
+	}
+	if s.base.lastHeight == unknownHeight || block.Height > s.base.lastHeight {
+		s.base.lastHeight = block.Height
+	}
+
+	return nil
+}
+
+func (s BlockStorage) LatestHeight() (uint64, error) {
+	s.base.mu.RLock()
+	defer s.base.mu.RUnlock()
+
+	if s.base.lastHeight == unknownHeight {
+		return 0, errors.NotInitialized
+	}
+	return s.base.lastHeight, nil
+}
+
+func (s BlockStorage) FirstHeight() (uint64, error) {
+	s.base.mu.RLock()
+	defer s.base.mu.RUnlock()
+
+	if s.base.firstHeight == unknownHeight {
+		return 0, errors.NotInitialized
+	}
+	return s.base.firstHeight, nil
+}
+
+var _ storage.ReceiptIndexer = &ReceiptStorage{}
+
+type ReceiptStorage struct {
+	base *baseStorage
+}
+
+func NewReceiptStorage() *ReceiptStorage {
+	return &ReceiptStorage{
+		base: baseStorageFactory(),
+	}
+}
+
+func (r ReceiptStorage) Store(receipt *gethTypes.Receipt) error {
+	r.base.mu.Lock()
+	defer r.base.mu.Unlock()
+
+	if _, ok := r.base.receiptsTxIDs[receipt.TxHash]; ok {
+		return errors.Duplicate
+	}
+
+	r.base.receiptsTxIDs[receipt.TxHash] = receipt
+	r.base.receiptBlockIDTxIDs[receipt.BlockHash] = receipt.TxHash
+
+	if _, ok := r.base.bloomHeight[receipt.BlockNumber.Int64()]; ok {
+		return errors.Duplicate
```

Comment:
I think this would mean that there were 2 receipts with the same ID but different `BlockNumber`. That seems like an inconsistent state